### PR TITLE
[move-2024] Enable Move 2024 optimizations

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

The Move 2024 optimizations got disabled from alpha at some point. This re-enables them.

## Test Plan 

The test suite still works, plus more tests to ensure the new freeze mechanisms work.

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/editions/mod.rs [hunk 1]
```diff
@@ -118,6 +118,7 @@ const E2024_ALPHA_FEATURES: &[FeatureGate] = &[
     FeatureGate::BlockLabels,
     FeatureGate::Move2024Paths,
     FeatureGate::MacroFuns,
+    FeatureGate::Move2024Optimizations,
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/editions/mod.rs:121]: I'm confused. I don't see this used anywhere?

Also, are we ready to turn on these optimizations for all versions?

Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/editions/mod.rs:121]: The freeze change in HLIR is feature gated by this. I have updated the diff to also enable the CFGIR optimization. This was also partially to help @rvantonder verify the versioning work.

We can postpone if you'd like.

Comment (rvantonder) [reply external-crates/move/crates/move-compiler/src/editions/mod.rs:121]: Just to chip in, nonessential for this to go in now on my account (I can test locally with patches). Deferring to yinz for making sure this is ready™️

Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/editions/mod.rs:121]: I also. had it in mind that we would set up a `2024.prerelease` or some such edition for the things actually shipping in the prerelease.

**Review by tnowacki — APPROVED** [review]
It looks fine to me, I just hope we can soon remove any gating for optimizations :)

**Pull request merged**